File: src/api/federation/sync.py
import asyncio
import base64
import time

import httpx
import structlog

from api.federation.state import FederationStateManager
from config.provider import GlobalConfigProvider
from config.schema import FederationNodeState

logger = structlog.get_logger()

# ─────────────────────────────────────────────────────────────────────────────
# Circuit Breaker & Shared Client
# ─────────────────────────────────────────────────────────────────────────────

_FED_CLIENTS: dict = {}
_CIRCUIT_STATE: dict = {}
_FED_SECRET_CACHE: dict = {}


async def shutdown_fed_clients():
    for key, client in list(_FED_CLIENTS.items()):
        await client.aclose()
    _FED_CLIENTS.clear()
    _CIRCUIT_STATE.clear()


def _get_fed_client(trust_mode: str) -> httpx.AsyncClient:
    if trust_mode not in _FED_CLIENTS:
        limits = httpx.Limits(max_connections=10, max_keepalive_connections=5)
        _FED_CLIENTS[trust_mode] = httpx.AsyncClient(
            limits=limits,
            timeout=httpx.Timeout(5.0, connect=3.0),
            verify=(trust_mode == "verify"),
        )
    return _FED_CLIENTS[trust_mode]


def _build_fed_headers(srv_config) -> dict:
    node_id = srv_config.node_id
    if node_id not in _FED_SECRET_CACHE:
        _FED_SECRET_CACHE[node_id] = base64.b64encode(
            srv_config.secret.encode("utf-8")
        ).decode("utf-8")
    return {
        "X-Federation-Secret": _FED_SECRET_CACHE[node_id],
        "X-Federation-Node": node_id,
    }


def _circuit_is_open(node_id: str, threshold: int) -> bool:
    if node_id not in _CIRCUIT_STATE:
        return False
    return _CIRCUIT_STATE[node_id] >= threshold


def _circuit_record_failure(node_id: str) -> int:
    _CIRCUIT_STATE[node_id] = _CIRCUIT_STATE.get(node_id, 0) + 1
    return _CIRCUIT_STATE[node_id]


def _circuit_record_success(node_id: str):
    _CIRCUIT_STATE[node_id] = 0


# ─────────────────────────────────────────────────────────────────────────────
# Network Execution Modules
# ─────────────────────────────────────────────────────────────────────────────


async def _poll_node_with_circuit_breaker(
    node_id: str, state_mgr: FederationStateManager
):
    """Individual node poll with circuit breaker + retry + backoff."""
    config = GlobalConfigProvider().get_config()

    if node_id not in config.federation.server:
        return

    srv_config = config.federation.server[node_id]
    fed_config = config.federation

    # Check circuit breaker state
    if _circuit_is_open(node_id, fed_config.circuit_breaker_threshold):
        return  # Skip; will retry after backoff expires

    url = srv_config.url.rstrip("/")
    try:
        client = _get_fed_client(srv_config.trust_mode)

        # Parallel polling with per-node timeout
        resp = await client.get(
            f"{url}/health",
            headers=_build_fed_headers(srv_config),
            timeout=fed_config.per_node_timeout,
        )
        resp.raise_for_status()
        data = resp.json()

        await state_mgr.set_state(
            node_id,
            FederationNodeState(
                status="up",
                latency_ms=data.get("meta", {}).get("duration_ms", 0),
                last_check=time.time(),
                consecutive_failures=0,
                next_retry_at=0,
                databases=data.get("data", {}).get("checks", {}).get("databases", {}),
                storages=data.get("data", {}).get("checks", {}).get("storages", {}),
            ),
        )
        _circuit_record_success(node_id)

    except (httpx.TimeoutException, httpx.HTTPError) as e:
        failures = _circuit_record_failure(node_id)
        backoff = min(2**failures, fed_config.backoff_max)  # EXPONENTIAL BACKOFF

        logger.warning(
            "Failed to sync federated node",
            node_id=node_id,
            error=str(e),
            failures=failures,
            backoff=backoff,
        )

        await state_mgr.set_state(
            node_id,
            FederationNodeState(
                status="down"
                if failures >= fed_config.circuit_breaker_threshold
                else "degraded",
                latency_ms=0,
                last_check=time.time(),
                consecutive_failures=failures,
                next_retry_at=time.time() + backoff,
            ),
        )


# ─────────────────────────────────────────────────────────────────────────────
# Primary Daemon Task
# ─────────────────────────────────────────────────────────────────────────────


async def sync_federated_servers():
    """Background task polling federated servers securely ensuring network synchronization."""
    logger.info("Federation sync started")
    config = GlobalConfigProvider().get_config()

    if not config.features.federation or not config.federation.enabled:
        logger.info("Federation is disabled, shutting down sync task")
        return

    state_mgr = FederationStateManager()
    await state_mgr.load()

    # Seed new nodes
    for node_id in config.federation.server:
        if await state_mgr.get_state(node_id) is None:
            await state_mgr.set_state(node_id, FederationNodeState(status="unknown"))

    # Spawn listeners for grpc enabled nodes
    tasks = []
    http_nodes = []

    for node_id, srv_config in config.federation.server.items():
        if srv_config.grpc_enabled:
            tasks.append(
                asyncio.create_task(_subscribe_node_health(node_id, state_mgr))
            )
        else:
            http_nodes.append(node_id)

    # Poll HTTP nodes
    while True:
        try:
            config = GlobalConfigProvider().get_config()
            nodes_to_poll = await state_mgr.get_next_retry_nodes()

            # Add nodes that haven't been polled in a while based on sync interval
            now = time.time()
            for node_id in http_nodes:
                state = await state_mgr.get_state(node_id)
                if (
                    state
                    and state.status == "up"
                    and (now - state.last_check) >= config.federation.sync_interval
                ):
                    if node_id not in nodes_to_poll:
                        nodes_to_poll.append(node_id)

            if nodes_to_poll:
                await asyncio.gather(
                    *[
                        _poll_node_with_circuit_breaker(node_id, state_mgr)
                        for node_id in nodes_to_poll
                    ],
                    return_exceptions=True,
                )
                await state_mgr.persist()

            degraded = False
            for node_id in nodes_to_poll:
                st = await state_mgr.get_state(node_id)
                if st and st.status == "degraded":
                    degraded = True
                    break

            sleep_time = min(config.federation.sync_interval, 5 if degraded else 30)
            await asyncio.sleep(sleep_time)

        except asyncio.CancelledError:
            logger.info("Federation sync shutting down")
            for t in tasks:
                t.cancel()
            break
        except Exception as sync_exception:
            logger.error("Federation sync error", error=str(sync_exception))
            await asyncio.sleep(config.federation.sync_interval)


async def _subscribe_node_health(node_id: str, state_mgr: FederationStateManager):
    from api.federation.grpc_client import get_grpc_client

    while True:
        try:
            client = get_grpc_client(node_id)
            if not client:
                await asyncio.sleep(10)
                continue

            async for update in client.health_stream():
                # Convert proto NodeStatus to string
                status_str = "unknown"
                if update.status == 1:
                    status_str = "up"
                elif update.status == 2:
                    status_str = "degraded"
                elif update.status == 3:
                    status_str = "down"

                await state_mgr.set_state(
                    node_id,
                    FederationNodeState(
                        status=status_str,
                        latency_ms=update.latency_ms,
                        databases=dict(update.databases),
                        storages=dict(update.storages),
                        last_check=time.time(),
                    ),
                )
                await state_mgr.persist()

        except asyncio.CancelledError:
            break
        except Exception as e:
            logger.warning(
                "gRPC Health stream error, reconnecting...",
                node_id=node_id,
                error=str(e),
            )
            await asyncio.sleep(5)

File: src/api/federation/sync.rs
use std::time::{SystemTime, UNIX_EPOCH};

use dashmap::DashMap;
use once_cell::sync::Lazy;
use tracing::{error, info, warn};

use crate::api::federation::state::{FederationNodeState, FederationStateManager};
use crate::config::loader::ConfigManager;

fn now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

// Circuit-breaker failure counters per node
static CIRCUIT_STATE: Lazy<DashMap<String, u32>> = Lazy::new(DashMap::new);

fn circuit_is_open(node_id: &str, threshold: u32) -> bool {
    CIRCUIT_STATE
        .get(node_id)
        .map(|v| *v >= threshold)
        .unwrap_or(false)
}

fn circuit_record_failure(node_id: &str) -> u32 {
    let mut entry = CIRCUIT_STATE.entry(node_id.to_string()).or_insert(0);
    *entry += 1;
    *entry
}

fn circuit_record_success(node_id: &str) {
    CIRCUIT_STATE.insert(node_id.to_string(), 0);
}

async fn poll_node(node_id: &str, state_mgr: &FederationStateManager) {
    let config = ConfigManager::get();

    let fed_config = &config.federation;

    let srv_config = match fed_config.server.get(node_id) {
        Some(sc) => sc.clone(),
        None => return,
    };

    let threshold = fed_config.circuit_breaker_threshold as u32;
    if circuit_is_open(node_id, threshold) {
        return; // Circuit is open — skip this node until backoff expires
    }

    let url = srv_config.url.trim_end_matches('/');
    let health_url = format!("{}/health", url);

    use base64::Engine;
    let mut headers = reqwest::header::HeaderMap::new();
    let b64_secret = base64::engine::general_purpose::STANDARD.encode(&srv_config.secret);
    headers.insert(
        "X-Federation-Secret",
        reqwest::header::HeaderValue::from_str(&b64_secret).unwrap(),
    );
    headers.insert(
        "X-Federation-Node",
        reqwest::header::HeaderValue::from_str(&srv_config.node_id).unwrap(),
    );

    let timeout = std::time::Duration::from_secs_f64(fed_config.per_node_timeout);
    let start = std::time::Instant::now();

    let client = reqwest::Client::builder()
        .timeout(timeout)
        .danger_accept_invalid_certs(srv_config.trust_mode != "verify")
        .default_headers(headers)
        .build();

    match client {
        Ok(http) => match http.get(&health_url).send().await {
            Ok(resp) if resp.status().is_success() => {
                let latency_ms = start.elapsed().as_secs_f64() * 1000.0;
                let data: serde_json::Value = resp.json().await.unwrap_or_default();

                let databases = data
                    .get("data")
                    .and_then(|d| d.get("checks"))
                    .and_then(|c| c.get("databases"))
                    .and_then(|v| v.as_object())
                    .map(|m| m.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
                    .unwrap_or_default();

                let storages = data
                    .get("data")
                    .and_then(|d| d.get("checks"))
                    .and_then(|c| c.get("storages"))
                    .and_then(|v| v.as_object())
                    .map(|m| m.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
                    .unwrap_or_default();

                state_mgr.set_state(
                    node_id,
                    FederationNodeState {
                        status: "up".to_string(),
                        latency_ms,
                        last_check: now_secs(),
                        consecutive_failures: 0,
                        next_retry_at: 0.0,
                        databases,
                        storages,
                    },
                );
                circuit_record_success(node_id);
                info!(
                    "Federation node {} is up ({}ms)",
                    node_id, latency_ms as u64
                );
            }
            Ok(resp) => {
                let failures = circuit_record_failure(node_id);
                let backoff = (2u64.pow(failures)).min(fed_config.backoff_max as u64) as f64;
                warn!(
                    "Federation node {} returned HTTP {}",
                    node_id,
                    resp.status()
                );
                state_mgr.set_state(
                    node_id,
                    FederationNodeState {
                        status: if failures >= threshold {
                            "down"
                        } else {
                            "degraded"
                        }
                        .to_string(),
                        consecutive_failures: failures,
                        next_retry_at: now_secs() + backoff,
                        last_check: now_secs(),
                        ..Default::default()
                    },
                );
            }
            Err(e) => {
                let failures = circuit_record_failure(node_id);
                let backoff = (2u64.pow(failures)).min(fed_config.backoff_max as u64) as f64;
                warn!("Federation node {} unreachable: {}", node_id, e);
                state_mgr.set_state(
                    node_id,
                    FederationNodeState {
                        status: if failures >= threshold {
                            "down"
                        } else {
                            "degraded"
                        }
                        .to_string(),
                        consecutive_failures: failures,
                        next_retry_at: now_secs() + backoff,
                        last_check: now_secs(),
                        ..Default::default()
                    },
                );
            }
        },
        Err(e) => error!("Failed to build HTTP client for {}: {}", node_id, e),
    }
}

pub async fn sync_federated_servers() {
    info!("Federation sync daemon started");

    let config = ConfigManager::get();

    if !config.features.federation {
        info!("Federation disabled — sync daemon exiting");
        return;
    }
    let fed_cfg = &config.federation;

    let state_mgr = FederationStateManager::new();
    state_mgr.load().await;

    // Seed unknown states for all configured servers
    for node_id in fed_cfg.server.keys() {
        if state_mgr.get_state(node_id).is_none() {
            state_mgr.set_state(node_id, FederationNodeState::default());
        }
    }

    let sync_interval = fed_cfg.sync_interval as u64;

    loop {
        let config = ConfigManager::get();
        if !config.features.federation {
            break;
        }
        let fed_cfg = &config.federation;

        let mut nodes_to_poll = state_mgr.get_next_retry_nodes();

        // Add healthy nodes due for a sync refresh
        let now = now_secs();
        for node_id in fed_cfg.server.keys() {
            if let Some(state) = state_mgr.get_state(node_id) {
                if state.status == "up" && (now - state.last_check) >= sync_interval as f64 {
                    if !nodes_to_poll.contains(&node_id.to_string()) {
                        nodes_to_poll.push(node_id.to_string());
                    }
                }
            }
        }

        if !nodes_to_poll.is_empty() {
            let handles: Vec<_> = nodes_to_poll
                .iter()
                .map(|node_id| {
                    let node_id = node_id.clone();
                    let sm = FederationStateManager::new();
                    tokio::spawn(async move {
                        poll_node(&node_id, &sm).await;
                    })
                })
                .collect();

            for h in handles {
                let _ = h.await;
            }

            state_mgr.persist().await;
        }

        let has_degraded = nodes_to_poll.iter().any(|id| {
            state_mgr
                .get_state(id)
                .map(|s| s.status == "degraded")
                .unwrap_or(false)
        });

        let sleep_secs = if has_degraded { 5 } else { sync_interval };
        tokio::select! {
            _ = tokio::time::sleep(std::time::Duration::from_secs(sleep_secs)) => {}
            _ = tokio::signal::ctrl_c() => {
                info!("Federation sync shutting down");
                break;
            }
        }
    }
}
